File: 13数组的奇偶位置调整/Solution.cpp
class Solution {
public:
    void reOrderArray(vector<int> &array) {
        deque<int>  res;
        int len =array.size()-1;
        for (int i=0;i<=len;i++){
            if (array[len-i]%2==1) 
                res.push_front(array[len-i]);
            if(array[i]%2==0)
                res.push_back(array[i]);
        }
        array.assign(res.begin(),res.end());
    }
};
/* 使用双向队列，遍历数组，如果是奇数则头插，如果是偶数则尾插
但是需要注意的是，双指针中，奇数遍历指针要从数组尾部开始而
偶数遍历指针要从数组头部开始，这样才能保证奇偶数重排后的相对
位置不变 时间复杂度小而空间复杂度大*/

File: 13数组的奇偶位置调整/Solution.java

public class Solution {
    public void reOrderArray(int [] array) {
        int len=array.length-1;
        for (int i=len;i>0;i--){
            for(int j=0;j<i;j++){
                if(isEven(array[j])&&!isEven(array[j+1]))
                    swap(array,j,j+1);
            }
        }
    }
    private boolean isEven(int a){
        return a%2==0;
    }
    private void swap(int[] array,int a,int b){
        int t=array[a];
        array[a]=array[b];
        array[b]=t;
    }
}
/* 该解法运用的是冒泡的思维，使每一个偶数都和它的下一个数做比较，如果是奇数则
交换，这样就可以将所有的偶数都顺序冒泡到数组的后端，空间复杂度小而时间复杂度大 */